third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -463,574 +299,1153 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+                   ConversionPatternRewriter &rewriter) const override {
+    SmallVector<size_t> valRangeLens;
+    ArrayRef<ValueRange> remappedInits = adaptor.getInits();
+    for (ValueRange remappedInit : remappedInits)
+      valRangeLens.push_back(remappedInit.size());
+
+    // rewrite the "before" block bb args
+    unsigned inputNo = 0;
+    TypeConverter localTypeConverter;
+    localTypeConverter.addConversion(
+        [&inputNo, remappedInits](Type _inputType,
+                                  SmallVectorImpl<Type> &types) {
+          SmallVector<Type> remappedInitTypes =
+              llvm::to_vector(remappedInits[inputNo].getTypes());
+          types.append(remappedInitTypes);
+          inputNo++;
+          return success();
+        });
+    std::optional<TypeConverter::SignatureConversion> conversion =
+        localTypeConverter.convertBlockSignature(whileOp.getBeforeBody());
+    if (!conversion)
+      return failure();
+    auto newBeforeBodyBlock = rewriter.applySignatureConversion(
+        whileOp.getBeforeBody(), *conversion, &localTypeConverter);
+
+    auto propagateCanNarrowToBlock = [remappedInits, this](Block *block) {
+      int offset = 0;
+      for (auto operands : remappedInits) {
+        if (operands.size() == 2) {
+          fatPtrs[{block->getArgument(offset), block->getArgument(offset + 1)}]
+              .canNarrow = fatPtrs.at({operands[0], operands[1]}).canNarrow;
+        }
+        offset += operands.size();
+      }
+    };
+
+    // propagate canNarrow to bb arg fatPtrs in before bb
+    propagateCanNarrowToBlock(newBeforeBodyBlock);
+
+    // rewrite the "after" block bb args
+    conversion =
+        localTypeConverter.convertBlockSignature(whileOp.getAfterBody());
+    if (!conversion)
+      return failure();
+    auto newAfterBodyBlock = rewriter.applySignatureConversion(
+        whileOp.getAfterBody(), *conversion, &localTypeConverter);
+
+    // propagate canNarrow to bb arg fatPtrs in after bb
+    propagateCanNarrowToBlock(newAfterBodyBlock);
+
+    SmallVector<Value> initArgs = flattenValues(remappedInits);
+    SmallVector<Type> resultTypes = llvm::map_to_vector(
+        llvm::make_filter_range(
+            initArgs, [](Value v) { return !v.getType().isInteger(1); }),
+        [](Value v) { return v.getType(); });
+    auto newWhileOp =
+        rewriter.create<scf::WhileOp>(whileOp.getLoc(), resultTypes, initArgs);
+
+    newWhileOp->setAttrs(whileOp->getAttrs());
+    rewriter.inlineRegionBefore(whileOp.getBefore(), newWhileOp.getBefore(),
+                                newWhileOp.getBefore().end());
+    rewriter.inlineRegionBefore(whileOp.getAfter(), newWhileOp.getAfter(),
+                                newWhileOp.getAfter().end());
+
+    SmallVector<ValueRange> packedRets;
+    for (unsigned i = 0, offset = 0; i < valRangeLens.size(); i++) {
+      // skip %cond
+      if (remappedInits[i].size() == 1 &&
```

Comment:
Why this would skip only `%cond`?

Reply:
because we don't "remap" `%cond` ever (we never do `1:N` for that value)

Reply:
What is not clear to me is why we cannot handle this as we handle other non-ptr parameters. I mean I think you are referring to this:
```
    %6 = scf.while (%arg1 = %5, %arg2 = %cond) : (tensor<1024x!tt.ptr<f32>, #blocked>, i1) -> (tensor<1024x!tt.ptr<f32>, #blocked>) {
        scf.condition(%arg2) %arg1 : tensor<1024x!tt.ptr<f32>, #blocked>
    } do {
        ^bb0(%arg1: tensor<1024x!tt.ptr<f32>, #blocked>):
            scf.yield %arg1, %cond : tensor<1024x!tt.ptr<f32>, #blocked>, i1
    }
```
Why `cond` should be handled differently from other non-pointer parameters? `%5` is our pointer and it will be split in `{baseptr, offset}`, but the rest will be untouched (I think). Sorry in advance if the question if naive :)

Reply:
Sorry I don't understand the question;

>  `but the rest will be untouched` 

This `continue` is exactly how they are untouched (because they are skipped). Anything that doesn't have a `1:2` conversion (as we receive in `OpOneToNAdaptor` is skipped because the only expected conversions are `1:2` (so any other conversion is not expected and in fact should never be written in this pass). The extra check here `remappedInits[i].getType()[0].isInteger(1))` should actually be an `assert` because we expect the *only* other things in the `scf.while` is the condition.

Reply:
Sorry I was doing a bit of "peephole" commenting. The reason we skip cond here is because here we're packing the results and `scf.while` doesn't return/result the cond variable.